obspy/core/event/base.py

```diff
@@ -611,43 +667,73 @@ def get_referred_object(self):
         Will return None if no object could be found.
         """
```

Comment:
> Ok, good point here but I am not sure how to do this correctly. In order to make this thread-safe(ish) I think the lock would need to be on read_events, and Catalog.copy as well wouldn't it?

Yea - true thread safety with all references working for reading events would require that I think.

> If I add a lock on read_events and Catalog.copy wouldn't it drastically reduce the benefit you get from multi-threading for solving the small corner case of duplicate resource_ids? 

The performance will probably not deteriorate a lot (if at all) as the locking would just be around the actual event parsing routines. The downloading would still be threaded and this is likely the slow part. With CPython's GIL even most threaded things don't truly run in parallel (I/O or downloading is an exception).

> If we do create a decorator for thread locking, where should such a decorator live, in obspy.util.decorator?

Yes. I'm not sure if the effort can be justified though as we are currently just talking about the `obspy.core.event.catalog._read()` and the two copy methods - a `with` statement in these would suffice.

> I dont understand how the race condition is created with the ResourceIdentifier methods you mentioned, although there probably is one I am just not seeing, as this is getting rather complicated.

Locking the higher level function you proposed is much cleaner in any case. I'm too lazy now to try to re-follow my own thoughts about locking these three methods...I think it was related to the fact that all the class level dictionaries are not thread local and with all the required shuffling of object ids and references it might result in some non-deterministic race conditions. But as you said: its getting fairly complicated and I might be mistaken.

All in all this deals with a rare corner case but then its only 10 lines or so to do it properly so why not I guess?